### PR TITLE
Improve intermediate IR file management and add support for unified compiler   

### PR DESCRIPTION
**Pull request by mehrdad2m**
**Context:**
We need to enhance Catalyst's intermediate IR file management system to support the unified compiler path that integrates xDSL passes. 

**Description of the Change:**
1. New `changed` level for `keep_intermediate`: Adds a new option to save IR files only when the IR is actually modified by a pass, reducing unnecessary file generation.

2. Enhances IR file organization by creating subdirectories for each compilation stage to better organize intermediate files

3. We want the initial unoptimized IR to be the same regardless of whether we are using xdsl passes or not. Therefore we need to dump the initial IR before any xdsl pass is applied. Printing the IR from Python will result in the generic format IR since the python context does not have all the dialects registered. To resolve this, We run `quantum-opt` on the IR before saving to ensure consistent, pretty format output. This introduces an extra call to `quantum_opt` but only when `keep_intermediate` is True

4. Adds callback to save IR after each xDSL passes. Note that an another [PR](https://github.com/PennyLaneAI/catalyst/pull/1978) adds the support on the c++ front

5. Pipeline stage renaming/shuffling: Moved all the quantum compilation passes up to the first stage so all of the quantum-level passes are applied at the same stage. The only exception is `lower-gradients` which currently uses lower level ops. We also renamed the first compilation stage from `enforce-runtime-invariants-pipeline` to `quantum-compilation-pipeline` and the old `quantum-compilation-pipeline` to `gradient-lowering-pipeline`. 

**Benefits:**
Better IR file management. 
E.g. When using `keep_intermediate="changed"` or `keep_intermediate="pass"` with xDSL passes, the output directory structure looks like:

```
workspace/
├── 0_workflow.mlir                           # Initial IR before any transformations
├── 1_AfterQuantumCompilationPass.mlir             # Stage-level summary
├── 2_AfterHLOLoweringPass.mlir               # Stage-level summary
├── 3_AfterGradientLoweringPass.mlir        # Stage-level summary
├── 4_AfterBufferizationPass.mlir            # Stage-level summary
├── 5_AfterMLIRToLLVMDialect.mlir            # Stage-level summary
├── 6_llvm_ir.ll                         # Final LLVM IR
├── workflow.so                                # Compiled shared object
├── workflow.o                                 # Object file
│
├── 0_QuantumCompilationPass/                 # Pass-level details for UserTransformPass (Python interpreter)
│   ├── 1_remove-chained-self-inverse.mlir    
│   ├── 2_xdsl-merge-rotations.mlir          
..
│
├── 1_QuantumCompilationPass/                #  User pass will be shown here if no xdsl pass is provided (Python interpreter)
│   └── 1_InlineNestedModulePass.mlir  # Plus other passes that are in this stage
│
├── 2_HLOLoweringPass/                         
│   ├── 1_StablehloLegalizeToLinalgPass.mlir
...
│
├── 4_BufferizationPass/                       
│   ├── 1_OneShotBufferizePass.mlir
...
│
└── 5_MLIRToLLVMDialect/                      
    ├── 1_SCFToControlFlowPass.mlir
...
```

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-103720]

### CODE REVIEW

File: frontend/catalyst/compiler.py [hunk 1]
```diff
@@ -349,9 +350,9 @@ def _options_to_cli_flags(options):
     if options.keep_intermediate:
         extra_args += ["--keep-intermediate"]
 
-    if options.keep_intermediate >= KeepIntermediateLevel.PASS:
-        extra_args += ["--save-ir-after-each=pass"]
-        extra_args += ["--dump-module-scope"]
+    if options.keep_intermediate >= KeepIntermediateLevel.CHANGED:
+        level = "pass" if options.keep_intermediate >= KeepIntermediateLevel.PASS else "changed"
+        extra_args += [f"--save-ir-after-each={level}", "--dump-module-scope"]
```
Comment (paul0403) [inline frontend/catalyst/compiler.py]: Why the `>= KeepIntermediateLevel.PASS`? That is the max level right? :thinking: 

I think it would be more readable to just do 
```python
match options.keep_intermediate:
   case PASS:
      level = "pass"
   case CHANGED:
      level = "changed"
```

File: mlir/include/Driver/CompilerDriver.h [hunk 2]
```diff
@@ -94,16 +94,35 @@ struct CompilerOutput {
     std::string outIR;
     std::string diagnosticMessages;
     PipelineOutputs pipelineOutputs;
-    size_t pipelineCounter = 0;
+    size_t globalPipelineCounter = 0; // Counter for root-level pipeline summary files
+    size_t passCounter = 0;           // Counter for passes within a pipeline folder
+    std::string currentStage = ".";   // Current compilation stage subdirectory
```
Comment (paul0403) [inline mlir/include/Driver/CompilerDriver.h:99]: Is this variable necessary? :sweat_smile: I feel like having the `"."` literal inlined in the `std::filesystem` lines is more readable, but up to you

File: frontend/catalyst/compiler.py [hunk 3]
```diff
@@ -571,6 +572,46 @@ def is_using_python_compiler(self, mlir_module=None):
...
+            workspace: The workspace directory path
+
+        Returns:
+            Callable or None: The callback function if intermediate saving is enabled, None otherwise
+        """
+        if not (workspace and self.options.keep_intermediate >= KeepIntermediateLevel.CHANGED):
+            return None
+
+        # pylint: disable-next=import-outside-toplevel
+        from xdsl.printer import Printer
```
Comment (paul0403) [inline frontend/catalyst/compiler.py:603]: Can we update the name of this function to reflect that it is xdsl path only?

File: frontend/catalyst/compiler.py [hunk 4]
```diff
@@ -571,6 +572,46 @@ def is_using_python_compiler(self, mlir_module=None):
...
+
+        user_transform_dir = os.path.join(str(workspace), "0_UserTransformPass")
+        os.makedirs(user_transform_dir, exist_ok=True)
+
+        pass_counter = 1
+
+        def save_pass_ir(previous_pass, module, _next_pass=None, **_kwargs):
+            """Callback to save IR after each pass in python_compiler."""
+            nonlocal pass_counter
```
Comment (paul0403) [inline frontend/catalyst/compiler.py]: `nonlocal` always scares me, especially for these factory functions. Do we have a way to avoid that? :thinking:

File: mlir/include/Driver/CompilerDriver.h [hunk 5]
```diff
@@ -94,16 +94,35 @@ struct CompilerOutput {
...
+        return std::filesystem::path(currentStage) /
+               std::filesystem::path(std::to_string(this->passCounter++) + "_" + pipelineName)
+                   .replace_extension(ext);
+    };
+
+    // Gets the root-level pipeline summary file name
+    std::string nextPipelineSummaryFilename(std::string pipelineName, std::string ext = ".mlir")
+    {
+        return std::filesystem::path(std::to_string(this->globalPipelineCounter) + "_After" +
+                                     pipelineName)
```
Comment (paul0403) [inline mlir/include/Driver/CompilerDriver.h:115]: Clarifying that the intermediates are always "after" is definitely a nice thing to have! :+1: 

Just one thing though, I think the final llvm ir should still be called "6_llvm_ir.ll", not "6_after_llvm_ir.ll", that sounds pretty weird :sweat_smile:

File: requirements.txt [hunk 6]
```diff
@@ -35,3 +35,4 @@ nbmake
 pennylane-lightning-kokkos
 amazon-braket-pennylane-plugin>1.27.1
 xdsl
+xdsl-jax @ git+https://github.com/xdslproject/xdsl-jax.git@main
```
Comment (mlxd) [inline requirements.txt]: https://pypi.org/project/xdsl-jax/

Any reason we cannot use the above?

Comment (paul0403) [reply requirements.txt]: I think this line was added when they weren't on pypi

Comment (paul0403) [reply requirements.txt]: https://github.com/PennyLaneAI/catalyst/pull/2193

File: mlir/include/Driver/CompilerDriver.h (hunk 2 above)
Comment (mehrdad2m) [reply mlir/include/Driver/CompilerDriver.h:99]: Well it's mutable. It get updated during lowering by `setStage()`. The "." is just the initial "no stage set yet" value but becomes things like "1_PipelineName" as we progress through pipelines.

File: frontend/catalyst/compiler.py (hunk 3 above)
Comment (mehrdad2m) [reply frontend/catalyst/compiler.py:603]: https://github.com/PennyLaneAI/catalyst/pull/2186/commits/5aa77796e9a2ceb8c9fd2d5fad4c73b550b7edc2

File: frontend/catalyst/compiler.py (hunk 4 above)
Comment (mehrdad2m) [reply frontend/catalyst/compiler.py]: https://github.com/PennyLaneAI/catalyst/pull/2186/commits/11edae72b7346e222271d3fc5db8c91c73fff250

File: mlir/include/Driver/CompilerDriver.h (hunk 5 above)
Comment (mehrdad2m) [reply mlir/include/Driver/CompilerDriver.h:115]: https://github.com/PennyLaneAI/catalyst/pull/2186/commits/70e196a072fcfae8cd264eb94e562e5ed6363075

File: frontend/catalyst/compiler.py (hunk 1 above)
Comment (mehrdad2m) [reply frontend/catalyst/compiler.py]: https://github.com/PennyLaneAI/catalyst/pull/2186/commits/e5d5da1666036c22f876655c3279c202690e4131

Comment (mehrdad2m) [reply frontend/catalyst/compiler.py]: yes that's the max. For some reason I thought we are supporting larger numbers to fall back to `KeepIntermediateLevel.PASS` like `keep_intermeidate = 5`

File: frontend/catalyst/compiler.py [hunk 7]
```diff
@@ -582,6 +587,52 @@ def is_using_python_compiler(self, mlir_module=None):
 
         return True
 
+    def _create_xdsl_pass_save_callback(self, workspace):
```
Comment (mudit2812) [inline frontend/catalyst/compiler.py:590]: Does it make sense to decorate this function with `@functools.lru_cache`? Seems to me like we may call it multiple times with the same `workspace`. Let me know if my assumption is incorrect.

**Review by mudit2812 — COMMENTED** [review]
Looks good to me. You have a soft approval from me, but I think the real approvals should come from the compilation team.

File: frontend/catalyst/compiler.py (hunk 7 above)
Comment (mehrdad2m) [reply frontend/catalyst/compiler.py:590]: We are calling `_create_xdsl_pass_save_callback` only once. What is being called more is the `__call__` method of `SavePassIRCallback`. but since it uses and modifies an outer scope variable (the `counter`) I am not sure if caching it would work 🤔

File: frontend/catalyst/pipelines.py [hunk 8]
```diff
@@ -222,8 +227,12 @@ def get_enforce_runtime_invariants_stage(_options: CompileOptions) -> List[str]:
         # But qnodes targeting other backends may choose to lower
         # this into something else.
         "inline-nested-module",
+        "annotate-function",
```
Comment (dime10) [inline frontend/catalyst/pipelines.py]: I think this new structure looks good! The only thing I was wondering about was this pass, I think it might for gradients but I'm not sure. I couldn't find out yet how the annotations are being used, from what I remember it was for verification (e.g. not differentiating an mcm)

Comment (mehrdad2m) [reply frontend/catalyst/pipelines.py]: Yeah you are right. This pass basically annotate functions that can't have gradients with `hasInvalidGradientOp` attribute and then runs a verifier right after it. Then the verification methods in the `gradient` dialect detect the attribute and raise an error. The name of the pass is definitely very confusing :) 
I think it could also live next to the `lower-gradients` pass. I just moved it to `QuantumCompilation` stage cause I thought we will move everything there eventually. But maybe it make sense to leave it in the `GradientLowering` stage for now to keep the semantic consistency of the stage names?

Comment (dime10) [reply frontend/catalyst/pipelines.py]: Yeah I think keeping the gradients in their own stage is not necessarily a bad thing :)

So we can update the annotate-pass name and move it there

Comment (mehrdad2m) [reply frontend/catalyst/pipelines.py]: https://github.com/PennyLaneAI/catalyst/pull/2186/commits/58244ba0d9516844e9fca901a071ce07b3f5a91f
